### DISCUSSION TITLE
"Date" header not changing between requests

### DISCUSSION THREAD

**Discussion by arjwilliams** [Kludex/uvicorn#1610] [Potential Issue] [answered]
I am testing out some cache-control headers in my web app, but the auto-generated "Date" header seems to only be changing when a new connection is established, rather than with each new request, which is preventing the browser from caching new responses, as it sees the response as being older than it actually is.

I would expect that the "Date" header to change between requests, even when using an existing connection.

I am testing on Python 3.10.5, with Uvicorn 0.18.2.

Possible test case, which is currently failing for me:
```python
import asyncio

from email.utils import parsedate_to_datetime
from contextlib import asynccontextmanager

import httpx

from uvicorn import Config, Server

async def test_default_date_header_different():
    config = Config(app=app, loop="asyncio", limit_max_requests=2)
    async with run_server(config):
        async with httpx.AsyncClient() as client:
            response1 = await client.get("http://127.0.0.1:8000")
            await asyncio.sleep(2)
            response2 = await client.get("http://127.0.0.1:8000")

            response1_date = parsedate_to_datetime(response1.headers["date"])
            response2_date = parsedate_to_datetime(response2.headers["date"])
            assert(response2_date > response1_date)

async def app(scope, receive, send):
    assert scope["type"] == "http"
    await send({"type": "http.response.start", "status": 200, "headers": []})
    await send({"type": "http.response.body", "body": b"", "more_body": False})

@asynccontextmanager
async def run_server(config: Config):
    server = Server(config=config)
    cancel_handle = asyncio.ensure_future(server.serve())
    await asyncio.sleep(0.1)
    try:
        yield server
    finally:
        await server.shutdown()
        cancel_handle.cancel()

asyncio.run(test_date_header_different())
```

**Comment by iudeen** [comment]
I checked the behavior with hypercorn (Daphne, doesn't support date header)

Server Code

```python

from hypercorn import Config
from hypercorn.asyncio import serve

config = Config()
config.max_app_queue_size=2

asyncio.run(serve(app, config))
```
The `date` header is changing on each request.

```python
>response1_date = parsedate_to_datetime(response1.headers["date"])
>response2_date = parsedate_to_datetime(response2.headers["date"])
>print(response2_date, response1_date)

2022-08-30 18:29:09+00:00 2022-08-30 18:29:07+00:00
```

**Reply by iudeen** [reply]
On checking Hypercorn's [implementation](https://gitlab.com/pgjones/hypercorn/-/blob/main/src/hypercorn/config.py#L272), we see that date is being assigned on formation of response header. 

While on Unicorn, its assigned using server state, which I suppose is initiated on new connection. (Not sure)

**Reply by iudeen** [reply]
If we agree this is an issue to be fixed, here is a way to fix.

We take reference from Hypercorn and set default headers on every response. If we do this, we might not need the `on_tick` function to update default header on every second.

**Comment by Kludex** [comment, marked as answer]
This was solved already. I don't remember when, but on 0.20.0 it's already working.
